### PR TITLE
edtlib: add support for pattern properties 

### PR DESCRIPTION
**Pull request by gmarull**
It is often useful to define pattern-based properties when writing
generic binding files included by other clients. For example, all
regulator clients are expected to refer to associated regulators using
the `.*-supply` pattern (`vin-supply`, `vdd-supply`, etc.). Until now,
this was not possible and so each client was forced to define its own
properties, making this generalization more difficult and error-prone.

Similar functionality is offered in dt-schema, the bindings language
used by the Linux Kernel and some other projects by using
`patternProperties`. Because it seems that dt-schema in Zephyr
won't happen anytime soon, this patch tries to modify our own language
to offer the same feature.

A new binding property attribute is added: `pattern`. If set to `true`,
the property name will be treated as a regular expression. This allows
user to define an arbitrary number of properties following that pattern.

Example usage:

```yaml
 # base.yaml

properties:
  ...
  ".*-supply$":
    type: phandle;
    pattern: true
  ...
```

```yaml
 # vnd,sensor.yaml

compatible = "vnd,sensor";

include: base.yaml

properties:
  vnd,foo:
    type: int
```

```dts
/* devicetree */
sensors_power: sensors-power {
    compatible = "regulator-fixed";
    ...
};

sensor {
    compatible = "vnd,sensor";
    vin-supply = <&sensors_power>;
    vnd,foo = <123>;
};
```

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
Great idea. Thanks.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:74]: Apparently, `typing.Pattern` is deprecated. Can you use `re.Pattern` instead?

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1431]: Bug: `name` is overwritten in the for loop above.
```suggestion
                _err(f"pattern property '{prop_spec.name}' in {self.binding_path} "
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1420]: Why disallow this?

**Review by 57300 — COMMENTED** [review]
I like it. This can be used to express inherited bindings more faithfully, which might ease the transition to dt-schema.

Patterns also allow for this kind of construction, as an alternative to PR https://github.com/zephyrproject-rtos/zephyr/pull/56788:

```yaml
properties:
  interrupts:
    type: array

  interrupts-extended:
    type: compound

  "^interrupts(-extended)?$":
    pattern: true
    required: true
    type: compound
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1420]: because a pattern property is always required to be in place. regular properties with defaults can be obviated from devicetree.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: good catch, fixed

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: fixed

**Comment by gmarull** [comment]
note: one feature this is missing is the ability to require specific properties matching a pattern, e.g.

```taml
# in a generic file
".*-supply$":
  type: phandle
  pattern: true

# in a specific binding file
vin-supply:
  required: true
```

adding DNM while I look into this

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1420]: I see. It took me a while to realize that it wouldn't even be possible to generate defaults for pattern properties.

**Comment by gmarull** [comment]
should now be ready, usages can be found e.g.

https://github.com/zephyrproject-rtos/zephyr/pull/63140
https://github.com/zephyrproject-rtos/zephyr/pull/63141

**Review by carlocaione — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]
LGTM

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:256]: I think the comments should be swapped, and `pattern_prop` shouldn't override `regular_prop` with empty values.
```suggestion
                    # overridable properties
                    if pattern_prop.type:
                        regular_prop.type = pattern_prop.type
                    if pattern_prop.deprecated:
                        regular_prop.deprecated = pattern_prop.deprecated
                    if pattern_prop.specifier_space:
                        regular_prop.specifier_space = pattern_prop.specifier_space
                    # non-overridable properties
                    if regular_prop.enum is None:
                        regular_prop.enum = pattern_prop.enum
                    if regular_prop.const is None:
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:555]: Duplicated sentence.
```suggestion
      True if the property is deprecated; False otherwise. It can be inherited
      from a matching pattern property.
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:579]: Dropped types that used to be attached to property handlers.
```suggestion
        self.type: str = self._raw.get("type", "")
        self.enum: Optional[list] = self._raw.get("enum")
        self.const: Union[None, int, List[int], str, List[str]] = self._raw.get("const")
        self.deprecated: bool = self._raw.get("deprecated", False)
        self.required: bool = self._raw.get("required", False)
        self.specifier_space: Optional[str] = self._raw.get("specifier-space")
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1424]: Stale comment. This function now initializes more than one property.
It could also use a new name, like maybe `_init_props_from_spec()`.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1447]: It would be nice to combine this check with the one in `_prop_val()`.
At the very least, note the missing `and self.status == "okay"`.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:2624]: This seems a little fishy. We shouldn't ignore the checks below when the type hasn't been set yet.

Do you think we could instead move the `self._check()` in `Bindings.__init__()` further down, once all the types have been assigned?

**Review by 57300 — CHANGES_REQUESTED** [review]
Good stuff, just a couple of suggestions.

It's worth noting that the example in https://github.com/zephyrproject-rtos/zephyr/pull/62921#pullrequestreview-1637666674 no longer works, because making `"^interrupts(-extended)?$"` required now makes `interrupts` AND `interrupts-extended` required, instead of OR. But this is okay, and probably fits most situations.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:256]: btw, an overridable property is a property that, if defined in a matching property (e.g. `vin-supply` for `.*-supply$`) will override the values from the pattern property. non-overridable means that if set in the pattern property, it will be enforced.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1447]: _prop_val() checks a single matching prop, this check verifies that there are at least >= 1 matching props if the pattern property is marked as required. how is status okay related here?

**Comment by gmarull** [comment]
> Good stuff, just a couple of suggestions.
> 
> It's worth noting that the example in [#62921 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/62921#pullrequestreview-1637666674) no longer works, because making `"^interrupts(-extended)?$"` required now makes `interrupts` AND `interrupts-extended` required, instead of OR. But this is okay, and probably fits most situations.

the thing is that if a pattern is required, so should be any matching properties. I don't see any other way around this.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:2624]: let me check, agreed this is not ideal and can skip real problems that were detected before.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:256]: Okay, I thought it was the other way around, but I see that your view is reflected in the documentation.

For other dummies like me, maybe it would be more readable to reorder these items, and add more verbose comments:
```python
# Attributes taken from the pattern property
...

# Attributes constrained by the regular property
...
```

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1447]: Because required properties can be ignored for nodes with status != okay. I believe this is also the case in dt-schema.

**Comment by 57300** [comment]
> the thing is that if a pattern is required, so should be any matching properties. I don't see any other way around this.

To be clear, I'm not requesting any changes in this regard.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1447]: oh, right, thanks for clarifying

Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: fixed

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:2626]: still looking into this. to make things easier, what we should do is to do the pattern property merges (now done after lookup tables) early in the process.

**Review by galak — CHANGES_REQUESTED** [review]
I'm not in favor of this, I think it makes more sense for bindings to be explicit about their properties.

**Comment by gmarull** [comment]
> I'm not in favor of this, I think it makes more sense for bindings to be explicit about their properties.

Are you aware that dt-schema offers the same feature? How are we ever gonna migrate to it? Also, how can the supply case be solved properly?

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: should be fixed in latest revision, let's see if I haven't created more issues :-)

**Comment by gmarull** [comment]
Note: this patch will ease to specify supply dependencies properly in devicetree, that is, regulators attached to an arbitrary number of nodes. Currently, most (if not all) are not captured in devicetree, things "just work" because regulators have a lower init priority. However, as soon as we move to ordinals as a sorting mechanism, proper supply specification will be mandatory. As of today, we have "vin-supply" defined in power.yaml, included in base.yaml. However, this is not a proper solution (supply name is arbitrary), not to mention that it is not aligned with Linux binding, a requirement in Zephyr.

**Comment by henrikbrixandersen** [comment]
> Note: this patch will ease to specify supply dependencies properly in devicetree, that is, regulators attached to an arbitrary number of nodes. Currently, most (if not all) are not captured in devicetree, things "just work" because regulators have a lower init priority. However, as soon as we move to ordinals as a sorting mechanism, proper supply specification will be mandatory. As of today, we have "vin-supply" defined in power.yaml, included in base.yaml. However, this is not a proper solution (supply name is arbitrary), not to mention that it is not aligned with Linux bindings, a requirement in Zephyr.

Sure, being able to reflect which regulator is regulating the supply to which external peripheral requires proper use of phandle properties, which this allows for. But this cannot stand on it own, right? I mean, this only works as long as the regulators are turned on (not just initialised) before the device is used, meaning all of these regulators will now rely on `regulator-boot-on` or perhaps `regulator-always-on` being set?

A full solution would require that the driver for the external peripheral is updated to have knowledge and support for any regulators referred to by any of the `*-supply` properties and turns these on and off as needed by the device operation - at which point the DTS binding for the given peripheral will need updating anyway to put in the expected name of the `*-supply` property (e.g. `vcc-supply`).

What am I missing?

**Review by mbolivar-ampere — COMMENTED** [review]
@galak: I am surprised by your opposition. To me this seems like a good idea. Can you clarify why we wouldn't want this now, given that a move to dt-schema would give it to zephyr anyway?

**Comment by gmarull** [comment]
last push:
- relaxed type override (it is allowed to re-specify the type in a matching property, follows dt-schema)
- dropped specifier-space checks on pattern properties (not feasible to check using regexes)

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1509]:
```suggestion
                      f"{self.binding_path} is required, but no matching "
```

File: doc/build/dts/bindings-syntax.rst
Comment (57300) [inline doc/build/dts/bindings-syntax.rst:481]: nit: This could be updated, since `type` is a special case.

**Review by 57300 — COMMENTED** [review]
LGTM, just spotted one mistake.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (gmarull) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: fixed

**Review by 57300 — DISMISSED** [review]

File: doc/build/dts/bindings-syntax.rst
Comment (gmarull) [reply doc/build/dts/bindings-syntax.rst]: reworded this a bit, in the end, they are not overridable even if they can be re-specified again.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:378]: This can be followed up in another PR, but should this be an error in general, beyond pattern properties? Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L327-L328

Same with `required: false`:
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L252-L254
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L711-L713

cc @mbolivar-ampere

**Review by 57300 — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
Can we have it such that `pattern` is only used for setting `type` and requires a further specification of the pattern to actually match a property in the dts.

I say this as I don't want people going wild with `pattern` usage and having to police it all the time, like I was trying to do with `default` and than pretty much gave up.

Additionally, I think the "leaf" bindings should be as specific as possible (unless you guys tell me there is some case / example that it doesn't make sense).  I'm thinking the leaf binding should specify a finite set of properties that is allowable for the node.

**Comment by gmarull** [comment]
> Can we have it such that `pattern` is only used for setting `type` and requires a further specification of the pattern to actually match a property in the dts.
> 

Does this mean `type: pattern` or something else? Note that patterns can have any type, as in dt-schema, so I'm not sure if I follow.

> I say this as I don't want people going wild with `pattern` usage and having to police it all the time, like I was trying to do with `default` and than pretty much gave up.
> 
> Additionally, I think the "leaf" bindings should be as specific as possible (unless you guys tell me there is some case / example that it doesn't make sense). I'm thinking the leaf binding should specify a finite set of properties that is allowable for the node.

Can you provide some more details/examples here?

**Comment by galak** [comment]
> > Can we have it such that `pattern` is only used for setting `type` and requires a further specification of the pattern to actually match a property in the dts.
> 
> Does this mean `type: pattern` or something else? Note that patterns can have any type, as in dt-schema, so I'm not sure if I follow.

I'm saying that if we specify `pattern: true` the attributes this applies to is `type:`.  So this is valid:

```
   properties:
     ".*-supply$":
        type: phandle
        pattern: true
```

this would report an error:

```
   properties:
     ".*-supply$":
        type: phandle
        pattern: true
        description: foobar
```



> 
> > I say this as I don't want people going wild with `pattern` usage and having to police it all the time, like I was trying to do with `default` and than pretty much gave up.
> > Additionally, I think the "leaf" bindings should be as specific as possible (unless you guys tell me there is some case / example that it doesn't make sense). I'm thinking the leaf binding should specify a finite set of properties that is allowable for the node.
> 
> Can you provide some more details/examples here?

1. Want to limit `pattern` usage to yaml files that will be included. (pretty much usage in dts/bindings/base/ only)
2. for my-device-binding.yaml:

I have to specify:
```
   properties:
     # matches '.*-supply$' from base.yaml, will be required.
     foo-supply:
       required: false
     bar-supply:
       required: true
```

For the `*-supply` properties to show up in the generated output.  Thus the my-device-binding.yaml is clearily specifying the specific `*-supply` properties that are valid for `my-device`.

**Comment by gmarull** [comment]
> > > Can we have it such that `pattern` is only used for setting `type` and requires a further specification of the pattern to actually match a property in the dts.
> > 
> > 
> > Does this mean `type: pattern` or something else? Note that patterns can have any type, as in dt-schema, so I'm not sure if I follow.
> 
> I'm saying that if we specify `pattern: true` the attributes this applies to is `type:`. So this is valid:
> 
> ```
>    properties:
>      ".*-supply$":
>         type: phandle
>         pattern: true
> ```
> 
> this would report an error:
> 
> ```
>    properties:
>      ".*-supply$":
>         type: phandle
>         pattern: true
>         description: foobar
> ```
> 

Why do we want this to report an error? What is the problem of adding descriptions to pattern properties?

> > > I say this as I don't want people going wild with `pattern` usage and having to police it all the time, like I was trying to do with `default` and than pretty much gave up.

I think pattern properties have nothing to do with the defaults problem. Defaults need to be justified, something that can't be enforced automatically, so here's where we have the problem. Pattern properties misusages can only be enforced via reviews, there's not much we can do if they are misused.

> > > Additionally, I think the "leaf" bindings should be as specific as possible (unless you guys tell me there is some case / example that it doesn't make sense). I'm thinking the leaf binding should specify a finite set of properties that is allowable for the node.
> > 
> > 
> > Can you provide some more details/examples here?
> 
> 1. Want to limit `pattern` usage to yaml files that will be included. (pretty much usage in dts/bindings/base/ only)
> 2. for my-device-binding.yaml:
> 
> I have to specify:
> 
> ```
>    properties:
>      # matches '.*-supply$' from base.yaml, will be required.
>      foo-supply:
>        required: false
>      bar-supply:
>        required: true
> ```
> 
> For the `*-supply` properties to show up in the generated output. Thus the my-device-binding.yaml is clearily specifying the specific `*-supply` properties that are valid for `my-device`.

Sounds like a self-imposed limitation that does not align with dt-schema. There are numerous example of pattern properties in Linux bindings out of "base" bindings that do not require anything else. See for example

https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/arm/arm,vexpress-juno.yaml
https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/arm/apple/apple,pmgr.yaml
etc.

I've counted ~1K occurrences of the patternProperties node in the latest Linus tree. So what is the rationale of such rule?

**Comment by carlescufi** [comment]
Architecture WG:

- @galak states that bindings are specifications and having globbing may make this specification less clear. Hence, using patterns should be restricted to binding files that are included, and not "leave nodes".
- @henrikbrixandersen tends to agree with @galak, describing that having pattern matching in the leave nodes could lead to some of those properties being ignored
- @gmarull mentions that in order to switch to ordinals we will need to have the dependencies correctly specified
- @gmarull has looked in-depth into the Linux case where pattern matching is allowed everywhere. When we switch to dtschema we will need to align with it in general
- @galak comments about the dtschema. It's unlikely that we'll get there and even then some of the choices in the dtschema may not have been the most appropriate ones. Among the things that we will need to align if we ever move to dtschema this would be another one.

**Comment by gmarull** [comment]
> * @galak comments about the dtschema. It's unlikely that we'll get there and even then some of the choices in the dtschema may not have been the most appropriate ones. Among the things that we will need to align if we ever move to dtschema this would be another one.

TBH, I think it's really Zephyr that created a language inferior to dt-schema. I guess Linux guys understood that creating new languages is a bad business, so that's why they chose a standard: json-schema. Solely with the amount of compile-time stuff we'd be able to catch with dt-schema (node naming, conditionals, etc.) we would improve user experience quite a lot. I'm surprised by this pushback, I always understood that our current language was seen as an historical mistake.

I'm fine closing this PR if there's not enough interest in this by the maintainers. I just feel we're loosing an opportunity to improve things just because of bike shedding. And... this time changes are non-breaking.

**Comment by 57300** [comment]
I think what @galak suggested has actually been done in dt-schema. If you look at this example binding:
https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/leds/kinetic,ktd2692.yaml

it declares `vin-supply` among properties, and sets `additionalProperties: false`. This makes it so that a node may contain a `vin-supply`, but not `any-other-supply`. This works despite `.*-supply$` being defined in the core schema (with phandle type) because of how multiple schemas are combined.

Similarly, here's one of a few bindings that restricts `pinctrl-[0-9]+` to only `pinctrl-0` and `pinctrl-1`:
https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/display/mediatek/mediatek,dpi.yaml

So, this kind of limitation wouldn't be a new thing. That's not to say that every pattern property should be limited this way, but maybe it would be a good enough starting point.

When it comes to Linux alignment as a priority, there's already a problem with the "imported" core and Linux bindings being misused a lot in Zephyr, because the binding language is not powerful enough to express them as intended. Last time I checked, `dt-validate` raised warnings on almost every `BOARD.dts` file in tree (all but one, in fact). Personally, I was interested in this PR being something of a stepping stone towards rectifying that, regardless of whether Zephyr will adopt dt-schema in the end.

**Comment by gmarull** [comment]
> I think what @galak suggested has actually been done in dt-schema. If you look at this example binding: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/leds/kinetic,ktd2692.yaml
> 
> it declares `vin-supply` among properties, and sets `additionalProperties: false`. This makes it so that a node may contain a `vin-supply`, but not `any-other-supply`. This works despite `.*-supply$` being defined in the core schema (with phandle type) because of how multiple schemas are combined.
> 

I'm aware of additionalProperties, it's a good way to restrict usage of patterns. It must be added to every binding to restrict by default (the behavior that seems to be preferred), so that would have some tree-wide implications. I'll take a look if this is a wanted feature, but I'll wait for feedback first to avoid wasting my time.

**Comment by gmarull** [comment]
> When it comes to Linux alignment as a priority, there's already a problem with the "imported" core and Linux bindings being misused a lot in Zephyr, because the binding language is not powerful enough to express them as intended. Last time I checked, `dt-validate` raised warnings on almost every `BOARD.dts` file in tree (all but one, in fact). Personally, I was interested in this PR being something of a stepping stone towards rectifying that, regardless of whether Zephyr will adopt dt-schema in the end.

I think we've already deviated too much from Linux that moving to dt-schema is now a daunting task, I think we just don't have the man power to do that transition. I also feel like the amount of breakage would be really high.

Not using what the de-facto devicetree maintainers chose is also a problem in the sense we'll likely never have influence on any of their decisions.

**Pull request closed without merging**
